File: app-back/Services/MeetingService.cs
using CelulaApp.DTOs;
using CelulaApp.Models;
using CelulaApp.Repositories.Interfaces;
using CelulaApp.Services.Interfaces;

namespace CelulaApp.Services;

public class MeetingService : IMeetingService
{
    private readonly IMeetingRepository _meetingRepo;
    private readonly IStudyRepository _studyRepo;

    public MeetingService(IMeetingRepository meetingRepo, IStudyRepository studyRepo)
    {
        _meetingRepo = meetingRepo;
        _studyRepo = studyRepo;
    }

    public async Task<List<MeetingResponse>> GetAllAsync()
        => (await _meetingRepo.GetAllAsync())
            .Select(m => new MeetingResponse(m.Id, m.Date, m.StudyId, m.Study.Title, m.LastDrawnMemberId))
            .ToList();

    public async Task<MeetingDetailResponse?> GetByIdAsync(int id)
    {
        var meeting = await _meetingRepo.GetByIdWithDetailsAsync(id);
        if (meeting is null) return null;

        var participations = meeting.Participations
            .Select(p => new ParticipationResponse(p.MemberId, p.Member.Name, p.TimesSelected, p.TimesSpoken))
            .ToList();

        return new MeetingDetailResponse(
            meeting.Id, meeting.Date, meeting.StudyId,
            meeting.Study.Title, meeting.LastDrawnMemberId, participations
        );
    }

    public async Task<MeetingResponse> CreateAsync(MeetingRequest request)
    {
        var study = await _studyRepo.GetByIdAsync(request.StudyId)
            ?? throw new KeyNotFoundException($"Estudo com id {request.StudyId} não encontrado.");

        var meeting = new Meeting { Date = request.Date, StudyId = request.StudyId };
        var created = await _meetingRepo.AddAsync(meeting);

        return new MeetingResponse(created.Id, created.Date, created.StudyId, study.Title, null);
    }
}

File: app-backend/src/services/meetingService.js
const meetingRepo = require('../repositories/meetingRepository');
const studyRepo = require('../repositories/studyRepository');

function toResponse(m) {
  return {
    id: m.id,
    date: m.date,
    studyId: m.studyId,
    studyTitle: m.study?.title ?? '',
    lastDrawnMemberId: m.lastDrawnMemberId,
  };
}

async function getAll() {
  return (await meetingRepo.getAll()).map(toResponse);
}

async function getById(id) {
  const m = await meetingRepo.getByIdWithDetails(id);
  if (!m) return null;

  return {
    id: m.id,
    date: m.date,
    studyId: m.studyId,
    studyTitle: m.study?.title ?? '',
    lastDrawnMemberId: m.lastDrawnMemberId,
    participations: m.participations.map((p) => ({
      memberId: p.memberId,
      memberName: p.member.name,
      timesSelected: p.timesSelected,
      timesSpoken: p.timesSpoken,
    })),
  };
}

async function create(body) {
  const study = await studyRepo.getById(body.studyId);
  if (!study) throw Object.assign(new Error(`Estudo com id ${body.studyId} não encontrado.`), { status: 404 });

  const m = await meetingRepo.create({ date: new Date(body.date), studyId: body.studyId });
  return toResponse(m);
}

async function remove(id) {
  return meetingRepo.remove(id);
}

module.exports = { getAll, getById, create, remove };
